### ISSUE TITLE
Deconvolution - performance regression

### ISSUE THREAD

**Issue by w-adamski** [XiaoMi/mace#210]
### System information
- **OS Platform and Distribution**: BlackBerry KEY2 (android running on SDM660 and using only 4 faster cores)
- **NDK, GCC, ...**: from docker: REPOSITORY: registry.cn-hangzhou.aliyuncs.com/xiaomimace/mace-dev IMAGE: 3e13cb346e2e

### Describe the problem
After commit c2523ee2ace8ce846b6279b0680a52a1050022b8 the library may be a lot slower for models using deconvolution because the computation may not be parallelized. The new implementation parallelize based on in_shape[0] (I assume batch size) and out_shape[1] (I assume number of output channels). For inference we can safely assume that batch size may be 1. For some networks number of output channels may be high, however for a network with a single doconvolution layer at the end of the network there will probably be at most 3 channels and often only one. So for such models the new implementation makes the inference slower than the previous one. And the previous one can be made a lot faster. Below is [not finished yet, but all tests pass] implementation based on the previous version that is faster than previous implementation and also approximatelly 20% faster than the new one after after making it parallel (most likely incorrectly since I've simple moved down pragma without checking for possible concurrent access):

```template<typename T>
void Deconv2dNCHW(const T *input,
                  const T *filter,
                  const T *bias,
                  const index_t *in_shape,
                  const index_t *out_shape,
                  const index_t *kernel_hw,
                  const int *strides,
                  const int *padding,
                  float *output) {

  const index_t in_shape_1 = in_shape[1];
  const index_t in_shape_2 = in_shape[2];
  const index_t in_shape_3 = in_shape[3];
  const index_t out_shape_0 = out_shape[0];
  const index_t out_shape_1 = out_shape[1];
  const index_t out_shape_2 = out_shape[2];
  const index_t out_shape_3 = out_shape[3];
  const index_t kernel_hw_0 = kernel_hw[0];
  const index_t kernel_hw_1 = kernel_hw[1];
  const int strides_0 = strides[0];
  const int strides_1 = strides[1];
  const int padding_0 = padding[0];
  const int padding_1 = padding[1];
  const index_t strides_0_kernel_hw_1 = strides_0 * kernel_hw_1;

#pragma omp parallel for collapse(4)
  for (index_t b = 0; b < out_shape_0; ++b) {
    for (index_t oc = 0; oc < out_shape_1; ++oc) {
      for (index_t oh = 0; oh < out_shape_2; ++oh) {
        for (index_t ow = 0; ow < out_shape_3; ++ow) {
          const index_t b_in_shape_1 = b * in_shape_1;
          const index_t oc_in_shape_1 = oc * in_shape_1;
          const index_t start_x = std::max<int>(0, ow + strides_1 -1 - padding_1) / strides_1;
          const index_t start_y = std::max<int>(0, oh + strides_0 -1 - padding_0) / strides_0;
          const index_t filter_start_x = kernel_hw_1 - 1 - (padding_1 + strides_1 * start_x - ow);
          const index_t filter_start_y = kernel_hw_0 - 1 - (padding_0 + strides_0 * start_y - oh);
          const index_t out_pos = ((b * out_shape_1 + oc) * out_shape_2 + oh) * out_shape_3 + ow;
          const int x_count = std::min(filter_start_x / strides_1 + 1, in_shape_3 - start_x);
          const int y_count = std::min(filter_start_y / strides_0 + 1, in_shape_2 - start_y);
          T out_value = 0;

          for (index_t ic = 0; ic < in_shape_1; ++ic) {
            const T * p_input_y = input + (b_in_shape_1 + ic) * in_shape_2 * in_shape_3 + start_x + start_y * in_shape_3;
            const T * p_filter_y = filter + (oc_in_shape_1 + ic) * kernel_hw_0 * kernel_hw_1 + filter_start_x + filter_start_y * kernel_hw_1;

            for (
                const T * p_input_y_end = p_input_y + y_count * in_shape_3;
                p_input_y < p_input_y_end;
                p_input_y += in_shape_3, p_filter_y -= strides_0_kernel_hw_1
            ) {
              const T * p_input = p_input_y;
              const T * p_filter = p_filter_y;

              for (const T * p_input_end = p_input + x_count; p_input < p_input_end; ++p_input, p_filter -= strides_1) {
                out_value += *p_input * *p_filter;
              }
            }
          }

          if (bias != nullptr)
            out_value += bias[oc];
          output[out_pos] = out_value;
        }
      }
    }
  }
}
```

### To Reproduce
The model parameters I've tested:
in_shape[0]: 1
in_shape[1]: 32
in_shape[2]: 1014
in_shape[3]: 762
out_shape[0]: 1
out_shape[1]: 1
out_shape[2]: 2035
out_shape[3]: 1531

**Comment by TonyMou** [comment]
@w-adamski thanks for your sharing. We will check this problem. BTW, which kernel size did you use?

**Comment by w-adamski** [comment]
kernel: 9x9
stride: 2x2

**Comment by TonyMou** [comment]
Fixed in commit 5addd5b778cff0d9cc09a496fcbd12fdba675b8d. @w-adamski feel free to reopen this issue.
